File: Week 5/solutions/G/G_opensource.cpp
#include <bits/stdc++.h>
using namespace std;

vector<pair<string,int>> opensource(vector<string>& lines){
    vector<pair<string,int>> ans;
    // We'll use this to store the users for each projects:
    unordered_map<string, unordered_set<string>> members;
    // We'll use this to store how many projects each user is in:
    unordered_map<string, int> user_projects;

    vector<string> projects = {lines[0]};
    string current = lines[0];
    for(int i = 1; i < lines.size(); i++){
        bool is_new_project = false;
        for(char c : lines[i]){
            is_new_project |= isupper(c);
        }
        if(is_new_project){
            current = lines[i];
            projects.push_back(current);
        }else if(!members[current].count(lines[i])){
            members[current].insert(lines[i]);
            user_projects[lines[i]]++;
        }
    }
    for(string s : projects){
        int ct = 0;
        for(string t : members[s]){
            ct += user_projects[t] == 1;
        }
        ans.push_back({s, ct});
    }
    sort(ans.begin(), ans.end(), [](pair<string,int> a, pair<string,int> b){
        if(a.second == b.second){
            return a.first < b.first;
        }
        return a.second > b.second;
    });
    return ans;
}

int main(){
    string s;
    vector<string> lines;
    while(true){
        getline(cin, s);
        if(s == "1"){
            auto ans = opensource(lines);
            for(auto p : ans){
                cout << p.first << ' ' << p.second << '\n';
            }
            lines.clear();
        }else if(s == "0"){
            break;
        }else{
            lines.push_back(s);
        }
    }
}

File: Week 5/solutions/G/G_opensource.py
def opensource(lines):
    ans = []
    # We'll use this to store the users for each projects:
    members = {}
    # We'll use this to store how many projects each user is in:
    user_projects = {}

    projects = [lines[0]]
    current = lines[0]
    members[current] = set()
    for i in range(1, len(lines)):
        is_new_project = False
        for c in lines[i]:
            is_new_project |= c.isupper()
        if is_new_project:
            current = lines[i]
            projects.append(current)
            members[current] = set()
        elif lines[i] not in members[current]:
            members[current].add(lines[i])
            user_projects[lines[i]] = user_projects.get(lines[i], 0) + 1
    for s in projects:
        ct = 0
        for t in members[s]:
            ct += user_projects[t] == 1
        ans.append((s, ct))
    ans.sort(key = lambda x: (-x[1], x[0]))
    return ans

lines = []
while True:
    s = input()
    if s == '1':
        ans = opensource(lines)
        for p in ans:
            print(p[0], p[1])
        lines = []
    elif s == '0':
        break
    else:
        lines.append(s)